### PR TITLE
Update Owner References strategy of DefaultPolicies and HelmReleases

### PR DESCRIPTION
**Pull request by matheusfm**
## Description 
Owner references with strange relations, causing infinite loops in create and deletion in default policies and some helm releases.

## Steps to reproduce the bugs

1. create a kind cluster: `KIND_CLUSTER_NAME=undistro ./hack/cluster.sh -k`
2. install undistro: `undistro --config config.yml install`
3. create a cluster: `undistro apply -f cluster.yaml` (just the cluster, without DefaultPolicies and Observer)
4. see logs from undistro controller manager pod: `kubectl logs -l undistro.io=undistro -n undistro-system -c manager --since 1s -f`
5. create the default policies referencing the cluster above: `kubectl apply -f defaultpolicies.yaml`

See that the reconciliation function is in loop, logging `Installing policy ...` from `controller.defaultpolicies`.

This loop is the **first problem** that can be fixed in another issue.

Let's see the tree of the resources:
`kubectl tree clusters.v1alpha1.app.undistro.io mycluster` see [kubectl-tree](https://github.com/ahmetb/kubectl-tree) plugin.
```
NAMESPACE  NAME                                                                        READY  REASON                   AGE
matheus    Cluster/mycluster                                                           True   ReconciliationSucceeded  69m
matheus    ├─Cluster/mycluster                                                         False  ScalingUp                69m
matheus    │ ├─AWSCluster/mycluster                                                    True                            69m
matheus    │ ├─AWSMachineTemplate/mycluster-cp-089e1ad2-2c66-4955-b499-147f4816da19    -                               69m
matheus    │ ├─KubeadmControlPlane/mycluster                                           False  ScalingUp                69m

ClusterAPI resources omitted ...

matheus    ├─HelmRelease/calico-mycluster                                              True   ReconciliationSucceeded  69m
matheus    └─HelmRelease/kyverno-mycluster                                             True   ReconciliationSucceeded  10m
```

Here, we can see that
1. there isn't DefaultPolicies CR in the tree;
2. the kyverno helm release is child of the cluster;
3. the kyverno helm release isn't child of DefaultPolicies, as the command below:
`kubectl get hr kyverno-mycluster -o yaml | yq e '.metadata.ownerReferences' -`
```yaml
- apiVersion: app.undistro.io/v1alpha1
  blockOwnerDeletion: true
  controller: true
  kind: Cluster
  name: mycluster
  uid: 59587fa9-784f-4d8f-bd3e-71be6af20676
```

Ok, let's delete the default policies with this command `kubectl delete -f defaultpolicies.yaml`.
See that the **terminal session is locked**:
```
defaultpolicies.app.undistro.io "defaultpolicies-mycluster" deleted

```

In other session we can see that the resource still exists:
`kubectl get defaultpolicies defaultpolicies-mycluster -o yaml | yq e '.metadata' -`
```yaml
annotations:
  kubectl.kubernetes.io/last-applied-configuration: |
    {"apiVersion":"app.undistro.io/v1alpha1","kind":"DefaultPolicies","metadata":{"annotations":{},"labels":{"undistro.io/move":""},"name":"defaultpolicies-mycluster","namespace":"matheus"},"spec":{"clusterName":"mycluster"}}
creationTimestamp: "2022-01-19T15:29:24Z"
deletionGracePeriodSeconds: 0
deletionTimestamp: "2022-01-19T15:54:05Z"
finalizers:
  - finalizer.undistro.io
generation: 2
labels:
  cluster.x-k8s.io/cluster-name: defaultpolicies-mycluster
  undistro.io: ""
  undistro.io/cluster-name: mycluster
  undistro.io/cluster-type: workload
  undistro.io/move: ""
name: defaultpolicies-mycluster
namespace: matheus
resourceVersion: "105354"
uid: f583ed26-e54b-441d-9579-9bc9f01d3cca
```

The resource contains `deletionTimestamp` and `finalizers` and it is the reason why it is not being deleted.

## Expected behavior

My expected behavior is that:
1. the DefaultPolicies to be deleted when I run `kubectl delete ...`.
2. the kyverno helm release to be cascaded deleted

My suggestion for this to happen is:
1. the DefaultPolicies CR should be owner of kyverno helm release
2. the cluster should not be owner of kyverno helm release (the Kubernetes garbage collector deletes objects that no longer have owner references)
3. when the DefaultPolicies CR is being deleted (`deletionTimestamp` is not null), it needs to remove the finalizer

## Who sets `OwnerReferences`?

| file                                              | function           | owner      | owned                     | condition                   | comments                                 |
|---------------------------------------------------|--------------------|------------|---------------------------|-----------------------------|------------------------------------------|
| `pkg/hr/helmrelease_client.go:44`                 | `Install`          | `Cluster`  | `HelmRelease`             | not be a management cluster | this client applies a HelmRelease object |
| `controllers/metadata/provider_controller.go:167` | `execMetadataFunc` | `Provider` | `Flavor` and `AWSMachine` | -                           | -                                        |
| `controllers/app/helmrelease_controller.go:131`   | `Reconcile`        | `Cluster`  | `HelmRelease`             | not be a management cluster | it uses a helm client to install a chart |
| `controllers/app/cluster_controller.go:214`       | `reconcile`        | `Cluster`  | Cluster API objects       | -                           | -                                        |

## Who uses `hr.Install()`?

| file                                                | function                         | chart                                                               |
|-----------------------------------------------------|----------------------------------|---------------------------------------------------------------------|
| `controllers/app/cluster_controller.go:537`         | `reconcileClusterAutoscaler`     | `cluster-autoscaler`                                                |
| `controllers/app/cluster_controller.go:485`         | `reconcileCNI`                   | `calico`                                                            |
| `controllers/app/defaultpolicies_controller.go:207` | `reconcile`                      | `kyverno`                                                           |
| `controllers/app/identity_controller.go:435`        | `reconcileComponentInstallation` | `pinniped-`                                                         |
| `controllers/app/observer_controller.go:513`        | `installRelease`                 | `kube-prometheus-stack`, `eck-operator`, `fluent-bit` and `fluentd` |
| `pkg/cloud/openstack/openstack.go:161`              | `ReconcileCloudProvider`         | `cloud-provider-openstack`                                          |

## Behavior after these changes

`kubectl tree clusters.v1alpha1.app.undistro.io mycluster` 
```
NAMESPACE  NAME                                                                        READY  REASON                   AGE  
matheus    Cluster/mycluster                                                           True   ReconciliationSucceeded  4h43m
matheus    ├─DefaultPolicies/defaultpolicies-mycluster                                 True   ReconciliationSucceeded  3m5s 
matheus    │ └─HelmRelease/kyverno-mycluster                                           True   ReconciliationSucceeded  3m4s 
matheus    └─HelmRelease/calico-mycluster                                              True   ReconciliationSucceeded  4h43m
```

Kyverno helm release has only `defaultpolicies-mycluster` as owner:
`kubectl get hr kyverno-mycluster -o yaml | yq e '.metadata.ownerReferences' -`
```yaml
- apiVersion: app.undistro.io/v1alpha1
  kind: DefaultPolicies
  name: defaultpolicies-mycluster
  uid: 6fd64afc-767b-49b1-adf5-c711b806ed91
```

If we delete default policy, the controller removes finalizer and the garbage collector of Kubernetes deletes the helm release.

```
kubectl delete -f defaultpolicies.yaml
defaultpolicies.app.undistro.io "defaultpolicies-mycluster" deleted
```

```
kubectl get hr
NAME               CLUSTER             CHART    VERSION   READY   STATUS                             AGE
calico-mycluster   matheus/mycluster   calico   3.19.1    True    Release reconciliation succeeded   4h50m
```

```
kubectl get defaultpolicies
No resources found in matheus namespace.
```
## Linked Issues
There isn't linked issues.

## Checklist
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/getupio-undistro/.github/labels?q=Type%3A)
- [x] I have documented my code (if applicable)
- [ ] My changes are covered by tests

### CODE REVIEW

**Review by carlostrr — APPROVED** [review]

File: pkg/hr/helmrelease_client.go [hunk 1]
```diff
@@ -34,15 +35,14 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
 )
 
-func Install(ctx context.Context, c client.Client, log logr.Logger, hr appv1alpha1.HelmRelease, cl *appv1alpha1.Cluster) error {
+func Install(ctx context.Context, c client.Client, log logr.Logger, hr appv1alpha1.HelmRelease, cl *appv1alpha1.Cluster, owners ...metav1.Object) error {
 	msg := fmt.Sprintf("Check condition for %s release", hr.Name)
```
Comment (hericlesme) [inline pkg/hr/helmrelease_client.go:39]:
```suggestion
```

File: pkg/hr/helmrelease_client.go [hunk 2]
```diff
@@ -34,15 +35,14 @@ import (
...
-	log.Info(msg, "lastAppliedVersion", hr.Status.LastAppliedRevision)
+	log.Info("Installing helm release", "helm release", hr.Name, "owners", owners, "lastAppliedVersion", hr.Status.LastAppliedRevision)
 	if meta.InReadyCondition(hr.Status.Conditions) {
 		meta.SetResourceCondition(cl, meta.ReadyCondition, metav1.ConditionTrue, meta.InstallSucceededReason, msg)
 	}
```
Comment (hericlesme) [inline pkg/hr/helmrelease_client.go:43]:
```suggestion
	if meta.InReadyCondition(hr.Status.Conditions) {
		msg := fmt.Sprintf("Check condition for %s release", hr.Name)
		meta.SetResourceCondition(cl, meta.ReadyCondition, metav1.ConditionTrue, meta.InstallSucceededReason, msg)
	}
```

**Comment by felipeweb** [comment]
Nice catch!!!!

File: pkg/hr/helmrelease_client.go (hunk 2 above)
Comment (matheusfm) [reply pkg/hr/helmrelease_client.go:43]: Thanks for the suggestion @hericlesme

**Review by hericlesme — APPROVED** [review]

**Comment by matheusfm** [comment]
I have fixed the loop problem described above in this commit https://github.com/getupio-undistro/undistro/pull/437/commits/fdfffe60c4124046b603cb5035eec9bb50557e42

The loop was happening because the CR was being edited (the sorting of `status.appliedPolicies` was always changing) in all reconciliation, triggering another one.

**Pull request merged**